src/background/providers/hcaptcha.ts

```diff
@@ -24,19 +81,179 @@ export class HcaptchaProvider implements Provider {
+            this.setStoredTokens(oldTokens.concat(newTokens));
+            this.forceUpdateIcon();
+        })();
+    }
+
     handleBeforeRequest(
         _details: chrome.webRequest.WebRequestBodyDetails,
     ): chrome.webRequest.BlockingResponse | void {
         return;
     }
+
     handleBeforeSendHeaders(
-        _details: chrome.webRequest.WebRequestHeadersDetails,
+        details: chrome.webRequest.WebRequestHeadersDetails,
     ): chrome.webRequest.BlockingResponse | void {
+        const url = new URL(details.url);
+        const urlType = this.handleUrl(url);
+
+        if (urlType.isIssuerUrl && details.method.toLowerCase() === 'post') {
+            // Erase any previous attempt
+            this.issueInfo = null;
+
+            // Do not store infinite tokens
+            const tokens = this.getStoredTokens();
+            if (tokens.length + NUM_OF_REQ_TOKENS > MAX_NUM_OF_TOKENS) return;
+
+            this.issueInfo = {
+                newUrl: details.url,
+                tabId: details.tabId,
+            };
+            return;
+        } else if (urlType.isSpendUrl) {
+            // Get one token
+            const tokens = this.getStoredTokens();
+            const oneToken = tokens.shift();
+            if (oneToken === undefined) return;
+            this.setStoredTokens(tokens);
+            this.forceUpdateIcon();
+
+            // Spend logic here
+            const httpPath = details.method + ' ' + url.pathname;
+            const binding = voprf.createRequestBinding(oneToken.getMacKey(), [
+                voprf.getBytesFromString(url.hostname),
+                voprf.getBytesFromString(httpPath),
+            ]);
+            const contents = [
+                voprf.getBase64FromBytes(oneToken.getInput()),
+                binding,
+                voprf.getBase64FromString(JSON.stringify(voprf.defaultECSettings)),
+            ];
+            const redemption = btoa(JSON.stringify({ type: 'Redeem', contents }));
+
+            const headers = details.requestHeaders ?? [];
+            headers.push({ name: 'challenge-bypass-token', value: redemption });
+            headers.push({ name: 'challenge-bypass-host', value: url.hostname });
+            headers.push({ name: 'challenge-bypass-path', value: httpPath });
+
+            this.issueInfo = null;
+            return {
+                requestHeaders: headers,
+            };
+        }
         return;
     }
+
     handleHeadersReceived(
-        _details: chrome.webRequest.WebResponseHeadersDetails,
+        details: chrome.webRequest.WebResponseHeadersDetails,
     ): chrome.webRequest.BlockingResponse | void {
+        if (details.responseHeaders === undefined) return;
+        if (this.issueInfo === null) return;
+        if (details.tabId !== this.issueInfo.tabId) return;
+
+        const url = new URL(details.url);
+        const urlType = this.handleUrl(url);
+        if (!urlType.isIssuerUrl || details.statusCode === 403) return;
```

Comment:
given preconditions are checked before calling `this.handleUrl()`, I think `details.statusCode === 403` should move at the same level.